### PR TITLE
Add BL parameter definitions as new document

### PR DESCRIPTION
**Pull request by theotherjimmy**
@AnotherButler, where should we link this from?

### CODE REVIEW

File: docs/tools/bootloader.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+For an guide on using the managed and unmanaged boot loader modes, see [the boot loader turorial](../tutorials/bootloder.md).
+
+There are 4 configuration parameters that affect the location of an application in ROM:
+ * `target.mbed_app_start`
+ * `target.mbed_app_size`
+ * `target.bootloader_img`
+ * `target.restrict_size`
+ 
+All of these parameters are valid in `targets.json`, `mbed_lib.json` and `mbed_app.json` and may be defined for individual targets and the wildcard target, `*`. A parameter in `mbed_lib.json` overrides a parameter in `targets.json`, and a parameter in `mbed_app.json` overrides one in `mbed_app.json`.
+
```
Comment (drahnr) [inline docs/tools/bootloader.md]: Is this intended? I assume one should be `mbed_lib.json`.

Comment (theotherjimmy) [reply docs/tools/bootloader.md]: Ha! Good catch. Yes, the second `mbed_app.json` should be `mbed_lib.json` in the last sentence.

Comment (theotherjimmy) [reply docs/tools/bootloader.md]: Fixed with a rebase.

File: docs/tools/bootloader.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+The start address of the current application, as computed above, is available to C and C++ as `APPLICATION_ADDR` and to the linker as `MBED_APP_START` ; The size of the current application is available as `APPLICATION_SIZE` and `MBED_APP_SIZE`. This parameter also defines `BOOTLOADER_ADDR` and `BOOTLOADER_SIZE` as the start address and size of the provided boot loader.
+
+This parameter may be used in conjunction with `target.restrict_size` and conflicts with `target.mbed_app_start` and `target.mbed_app_size`. When used with `target.restrict_size`, the size off the application is defined by that parameter; Otherwise, the size is the remainder of ROM.
+
+### `target.restrict_size`
+
+This parameter restricts the size of the application to be at most the specified size. When `target.bootloader_img` is present, the start of the current application's code segment is computed as above; Otherwise, the start address is the beginning of ROM. The size of the application computed by rounding the end address down to the nearest flash erase block boundary and subtracting the start address. The post-build merge process will pad the resulting boot loader binary to it's end address.
```
Comment (marcuschangarm) [inline docs/tools/bootloader.md]: I'm not completely sure what is the difference between `restrict_size` and `mbed_app_size`? It seems like you can use them for the same thing: define the size of your application?

I also don't understand the rounding down part: "The size of the application computed by rounding the end address down to the nearest flash erase block boundary and subtracting the start address"

File: docs/tools/bootloader.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+The start address of the current application, as computed above, is available to C and C++ as `APPLICATION_ADDR` and to the linker as `MBED_APP_START` ; The size of the current application is available as `APPLICATION_SIZE` and `MBED_APP_SIZE`. This parameter also defines `POST_APPLICATION_ADDR` and `POST_APPLICATION_SIZE` as the start address and size of the region after the application.
+
+This parameter may be used in conjunction with `target.bootloader_img` and conflicts with `target.mbed_app_start` and `target.mbed_app_size`. When used with `target.bootloader_img`, the start off the application is defined by that parameter; Otherwise, the start is the start of ROM.
+
+### Exporter limitations
+
+While the exporters can export a project using the configuration parameters above, there are some limitations. 
+
+The exporters do not interpret Mbed OS Configuration, and any changes to configuration parameters, especially boot loader parameter, require you to re-run the `mbed export` command.
```
Comment (marcuschangarm) [inline docs/tools/bootloader.md]: This problem is common to all configuration changes, right? It sounds like the bootloader is a special case.

Comment (theotherjimmy) [reply docs/tools/bootloader.md]: That's true. You could, in theory, modify `mbed_config.h` for other configuration parameters. BL configs don't even allow that.

File: docs/tools/bootloader.md (hunk 2 above)
Comment (theotherjimmy) [reply docs/tools/bootloader.md]: I'll have to think about how to phrase that better. The difference is that `restrict_size` will generate the `POST_APPLICATION_*` stuff and pad your executable to the size mentioned.

**Comment by marcuschangarm** [comment]
Much better! It was easy to understand and didn't leave any lingering questions behind!

File: docs/tutorials/bootloader.md [hunk 4]
```diff
@@ -99,12 +99,18 @@ For an example showing how to create an application that uses a bootloader, see
...
 
-The configuration option `target.mbed_app_size` defines the size of an application image in ROM by defining the `MBED_APP_SIZE` macro for the linker script. You may only define this configuration option on a per-target basis defined within the `target_overrides` section of an Mbed application configuration, and you may not define it for the metatarget `*`. When you do not define this configuration option, it defaults to the remaining ROM. The Mbed OS tools calculate the remaining ROM by subtracting the image's offset into ROM from the total size of ROM. Together with `target.mbed_app_start`, these configuration options define a continuous region of memory that an image may use. The tools verify that this region of memory is in ROM, but the tools do not perform any other checks for consistency or validity.
+Further, the exporters do not implement the postbuild merge that bootloader builds use. 
+
+For a managed bootloader:
+After exporting a project with the `target.mbed_app_start` setting, you are responsible for ensuring that a boot loader is present, if needed. Without flashing this boot loader image, the device will not boot correctly.
```
Comment (theotherjimmy) [inline docs/tutorials/bootloader.md]: This is an unmanaged BL parameter, and the unmanaged BL description. This should be swapped with the subsequent paragraph, without changing the title.

**Comment by theotherjimmy** [comment]
Much better.

**Comment by theotherjimmy** [comment]
Good to go!

**Pull request merged**